Add blog for KEP-5207: metrics.k8s.io/v1 GA (#56446)
kubernetes/website: content/en/blog/_posts/2026/kubernetes-v1-37-metrics-api-ga/index.md

---
layout: blog
title: "Kubernetes v1.37: Metrics API graduates to stable"
slug: kubernetes-v1-37-metrics-api-ga
draft: true
author: >
  [ChengHao Yang](https://github.com/tico88612)
---

Kubernetes v1.37 promotes the `metrics.k8s.io` API to stable (`v1`). This
API provides CPU and memory usage for nodes and Pods, and is the API behind
commands such as `kubectl top` and resource-metrics-based autoscaling.

For cluster operators and application developers, this graduation means that
the API now has the stability guarantees associated with a Kubernetes stable
API. The `v1` API has the same resource types and fields as `v1beta1`; this is
an API-version graduation, not a change to the metrics that are collected or
returned.

## A long-lived API reaches stable

The resource Metrics API was introduced as alpha in Kubernetes v1.6 and became
beta in v1.8. It has remained unchanged and has been used in production for
years by clients including the HorizontalPodAutoscaler (HPA) and `kubectl top`.
Kubernetes v1.37 formally graduates that proven API to `metrics.k8s.io/v1`.

The API exposes two resource types:

> - `NodeMetrics`, for CPU and memory usage for a node.

Reviewer: (nit) NodeMetrics, no backticks. The blog team can fix it up.

> - `PodMetrics`, for CPU and memory usage for a Pod, with a per-container

Reviewer: (nit) PodMetrics, no backticks. The blog team can fix it up.

>   breakdown through `ContainerMetrics`.

Reviewer: Is this correct:
```suggestion
- NodeMetrics, for CPU and memory usage for a node.
- PodMetrics, for CPU and memory usage for a Pod, including a per-container
  breakdown through `containerMetrics`.
```

I think PodMetrics and NodeMetrics are API types that you can fetch via HTTP, but ContainerMetrics is not.
(did I get that right?)

Author: Yes, that's right. `ContainerMetrics` is nested within the `containers` field of `PodMetrics`; it is not a separately fetchable API resource. I'll update the wording to make that distinction clear.

The API remains intentionally small. It provides the resource metrics needed
for autoscaling and basic inspection; it is not a replacement for a full

> monitoring pipeline or the Custom Metrics API.

Reviewer: (nit)
```suggestion
monitoring pipeline or the custom metrics (`custom.metrics.k8s.io`) API.
```
?

Author: Fixed.

> ## What changes in v1.37

Reviewer: ```suggestion
## What changed with the v1.37 release? {#changes}
```

Reviewer: (because this article will publish **after** `metrics.k8s.io` has graduated to stable).

Author: Thank you! Fixed.

The `v1` API surface is identical to `v1beta1`, except for the API version.
There are no renamed fields, new fields, or changes to the meaning of the
returned CPU and memory values.

For example, a client can retrieve node metrics from the stable endpoint:

```shell
kubectl get --raw /apis/metrics.k8s.io/v1/nodes
```

> Likewise, it can retrieve metrics for the Pods in a namespace:

Reviewer: (nit)
```suggestion
Likewise, it can retrieve metrics for the pods in a namespace:
```
(alternatively: also write Namespace in title case)

Author: Fixed.

```shell
kubectl get --raw /apis/metrics.k8s.io/v1/namespaces/default/pods
```

Kubernetes clients that support this graduation prefer `v1` when it is
available and fall back to `v1beta1` on clusters that do not yet serve `v1`.
This includes `kubectl top` and the in-tree resource metrics client used by
the HPA controller. The fallback keeps newer clients working with clusters

> that only provide `v1beta1`.

Reviewer: Is this true, that HPA supports the v1 API?

Author: `kubectl top` currently supports automatic fallback for `v1` and `v1beta1`, but HPA doesn't yet. Let me fix the description.

## What you need to do

No feature gate is required. However, the Metrics API is served through the
[API aggregation layer](/docs/tasks/extend-kubernetes/configure-aggregation-layer/)
by an implementation such as [metrics-server](https://github.com/kubernetes-sigs/metrics-server).
The implementation must register and serve the `v1.metrics.k8s.io` APIService

> before the stable endpoint is available in your cluster.

Reviewer: How about this?
```suggestion
You don't need to enable any feature gate. The Metrics API is served through the
[API aggregation layer](/docs/tasks/extend-kubernetes/configure-aggregation-layer/),
by an implementation such as [metrics-server](https://github.com/kubernetes-sigs/metrics-server).
You can choose any implementation of `metrics.k8s.io`; for the v1 metrics API to be
available in your cluster, your chosen implementation must serve the `v1.metrics.k8s.io` API, and you need to [register](/docs/tasks/extend-kubernetes/configure-aggregation-layer/) an associated [APIService](/docs/reference/kubernetes-api/apiregistration/api-service-v1/).
```

Author: The original wording seemed pretty harsh, but it's much better after these changes. Thanks.

During the transition, implementations should serve both `v1` and `v1beta1`.
Keeping both versions available maintains compatibility with older clients.
The `v1beta1` API remains available in Kubernetes v1.37.

You can see which versions your cluster serves with:

```shell
kubectl get --raw /apis/metrics.k8s.io/ | jq .
```

Once your metrics implementation supports `v1`, you can also check that its
APIService is available:

```shell
kubectl get apiservice v1.metrics.k8s.io
```

## Learn more

- Read the [Resource metrics pipeline](/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/)
  documentation.

> - Read [KEP-5207](https://kep.k8s.io/5207), the proposal for this graduation.

Reviewer: (nit)
```suggestion
- Read [KEP-5207](https://www.kubernetes.dev/resources/keps/5207/), the proposal for (graduating) this API.
```

Author: Fixed.

> - Learn about the [Metrics API](https://github.com/kubernetes/metrics#resource-metrics-api)

Reviewer: Ah, darn. If we're graduating it we should have documented it. Then we could hyperlink to that documentation

Did that get missed?

Reviewer: We can update the link if documentation gets merged, even after this PR is also merged.

Author: I was actually worried it might get blocked because I haven't updated the link yet, thanks!

  and its reference implementation, [metrics-server](https://github.com/kubernetes-sigs/metrics-server).

## Get involved

> The Metrics API is maintained by [SIG Instrumentation](https://github.com/kubernetes/community/tree/master/sig-instrumentation).

Reviewer: (nit)
```suggestion
The Metrics API is maintained by [SIG Instrumentation](https://www.kubernetes.dev/community/community-groups/sigs/instrumentation/).
```

Author: Fixed.
